### PR TITLE
[Bugfix] Preserve MXFP4 Triton weights in sharded state

### LINKED ISSUE #34448: [BUG] mxfp4 + sharded_state load format silently drops expert weights (gpt-oss-20b)
### Repro script

The script below loads `openai/gpt-oss-20b`, exports it with `save_sharded_state.py` (`--load-format sharded_state`), then compares the total size of the original checkpoint's safetensors against the exported ones. Run from the repo root:

```bash
#!/usr/bin/env bash
set -euo pipefail

MODEL_PATH="${1:?usage: $0 /path/to/openai/gpt-oss-20b [out_dir]}"
OUT_DIR="${2:-$(mktemp -d -t mxfp4-repro-XXXXXX)}"

python3 -c "import sglang; print('sglang version:', sglang.__version__)"
SGLANG_ROOT=$(python3 -c "import os, sglang; print(os.path.dirname(os.path.dirname(os.path.dirname(sglang.__file__))))")

python3 "${SGLANG_ROOT}/examples/runtime/engine/save_sharded_state.py" \
  --model-path "${MODEL_PATH}" \
  --trust-remote-code \
  --tensor-parallel-size 2 \
  --mem-fraction-static 0.7 \
  --attention-backend triton \
  --output "${OUT_DIR}"

echo "original checkpoint (safetensors only):"
du -ch "${MODEL_PATH}"/*.safetensors | tail -1
echo "sharded_state export (safetensors only):"
du -ch "${OUT_DIR}"/*.safetensors | tail -1
```

Run:

```
./repro_mxfp4_sharded_state_loss.sh /path/to/openai/gpt-oss-20b
```

### Result

Output (sglang 0.5.17, `lmsysorg/sglang:latest`, TP=2, GH200):

```
sglang version: 0.5.17
original checkpoint (safetensors only):
13G	total
sharded_state export (safetensors only):
4.0G	total
```

The export is missing every expert's packed FP4 weights (`w13_weight` / `w2_weight`); only the scales, biases, and everything else survive. Loading this checkpoint back with `--load-format sharded_state` runs every expert FFN on zeroed weights, with no error or warning.

### Problem source

This looks like a bug in sglang in the mxfp4 + `ShardedStateLoader` path.

**1. On quantization, the expert weights move off the module.** `Mxfp4MoEMethod.process_weights_after_loading` (triton_kernels branch, in `python/sglang/srt/layers/quantization/mxfp4.py`) moves the quantised weights to a helper object and then deletes them from `layer`, which is the `nn.Module`:

```python
self.w13_weight_triton_tensor = w13_weight
self.w2_weight_triton_tensor = w2_weight
del layer.w13_weight
del layer.w2_weight
```

**2. On save, `ShardedStateLoader.save_model` only sees what's still on the module.** It dumps `layer.state_dict()`, which by this point no longer contains `w13_weight` / `w2_weight`. They live on the quant-method helper object instead, which `state_dict()` never looks at.

**3. On load, the same deletion happens again, before the missing-key check can catch it.** `ShardedStateLoader.load_model` does:

```python
model = _initialize_model(...)
for _, module in model.named_modules():
    quant_method.process_weights_after_loading(module)
```

which zero-initializes the model before its missing-key check runs against it. The check is built from a `state_dict()` that's already missing `w13_weight` / `w2_weight`, so it never flags anything. The model loads "successfully" with those weights left at zero, with no error or warning.

### PR DESCRIPTION
**Pull request by jeremyzhang866**
## Summary

- keep Triton-swizzled MXFP4 expert weights and scales registered on the layer
- serialize non-contiguous postprocessed tensors through contiguous CPU snapshots
- keep wrapper-backed runtime tensors out of CPU/offload-group parameter remapping
- add CPU regressions for a fresh-layer sharded round trip and the offloader guard

## Root cause

The Triton MXFP4 post-load path moved the swizzled `w13` and `w2` weight wrappers onto the quantization method and deleted the registered layer parameters. `ShardedStateLoader.save_model` only exports `model.state_dict()`, so those runtime weights were omitted.

On reload, SGLang postprocesses freshly initialized placeholders before it builds the expected state dict. The missing weights therefore disappeared from both the checkpoint and the expected-key set, and loading silently succeeded with placeholder runtime weights. The swizzled scales held by `PrecisionConfig` had the same stale-storage problem even though the original scale parameters remained registered.

This change rebinds all four Triton runtime tensors (two weights and two scales) to their existing registered `Parameter` objects. The module state and kernel wrappers then share the same backing storage, so sharded-load copies update the tensors actually consumed by the kernels. Parameter identity and weight-loader attributes are preserved.

The Hopper swizzle can produce non-contiguous runtime tensors. Safetensors rejects those tensors, so the exporter materializes contiguous CPU snapshots at the save boundary without changing the live runtime layout.

## CPU offload compatibility

Triton wrappers live outside the module parameter mapping temporarily installed by `functional_call`. The four wrapper-backed Parameters are therefore marked to remain on device, and both offloader implementations honor that marker. This preserves the previous effective behavior of the hidden Triton tensors while allowing them to remain registered for serialization.

Fixes #34448

## Validation

- focused CPU regression with import-only stubs for unavailable GPU dependencies: `2 tests`, `OK`
  - creates source and zero-initialized destination layers through the real MXFP4 `create_weights`
  - runs the real MXFP4 post-load method with only the Triton layout conversion mocked
  - saves through the real `ShardedStateLoader`, reloads every safetensors entry into the fresh destination state, and verifies both weights and scales consumed by the runtime wrappers
  - verifies wrapper-backed Parameters are excluded from CPU offload
- the same sharded round-trip regression against current `upstream/main@b3bffef70a` fails as expected because `w13_weight` is absent after postprocessing
- `python3 scripts/ci/check_registered_tests.py`
- Black (`--check`) on all four changed files
- isort (`--check-only`) on the new test
- Ruff (`F401`, `F821`, `UP037`) on all four changed files
- `git diff --check`
- `python3 -m py_compile test/registered/unit/model_loader/test_mxfp4_sharded_state.py`

The CPU tests are registered in `base-a-test-cpu`.

- reporter validation on the original GH200, TP=2 reproduction: the sharded export increased from the broken `4.0G` result to `14G`, resolving the reported missing-weight export symptom ([result](https://github.com/sgl-project/sglang/pull/34558#issuecomment-5293077326))

A full save-reload logits comparison has not yet been reported. Project CI is still pending.

---
### CI States

Latest PR Test (Base): :x: [Run #32683915223](https://github.com/sgl-project/sglang/actions/runs/32683915223)
Latest PR Test (Extra): :x: [Run #32683915059](https://github.com/sgl-project/sglang/actions/runs/32683915059)
Latest PR Test (AMD ROCm 7.2): :x: [Run #32683915112](https://github.com/sgl-project/sglang/actions/runs/32683915112)

### CODE REVIEW

**Comment by youssef62** [comment]
Hi @jeremyzhang866, thanks for working on #34448. I ran your branch on the same repro script I added in that issue on a GH200 node, and it does solve the problem. 

Before fix: 
```
original checkpoint (safetensors only):
13G   total
sharded_state export (safetensors only):
4.0G  total
```

After: 
```
original checkpoint (safetensors only):
13G   total
sharded_state export (safetensors only):
14G   total
```

**Comment by jeremyzhang866** [comment]
Thanks @youssef62 for validating the original repro on GH200 and confirming that the sharded export returns from 4.0G to 14G.

@mmangkad, could you please review the MXFP4 runtime-state ownership and model-loader changes? @Fridge003, could you help shepherd this PR and, if the scope looks good, trigger `/tag-and-rerun-ci`?

The current red checks are gate failures caused only by the missing `run-ci` label: lint passed, and the actual test jobs have not run yet. The PR body now records the GH200 result while keeping the remaining full save-reload logits comparison explicit.

**Review by mmangkad — APPROVED** [review]

**Comment by mmangkad** [comment]
@jeremyzhang866 https://github.com/sgl-project/sglang/actions/runs/31862528898/job/94982279568 seems related

**Comment by youssef62** [comment]
@jeremyzhang866, @mmangkad I wonder if we could add a test that checks that all objects of type `QuantizeMethodBase` do not nullify a layer parameter to avoid(catch) an instance of this bug happening again(in other configs).

**Comment by jeremyzhang866** [comment]
CI attempt 2 still failed before the tests started on the 8-GPU B300 runner: [job log](https://github.com/sgl-project/sglang/actions/runs/31922959478/job/95124584150).

All 8 GPUs remained at about 86% memory usage, and the runner reported:

```
Orphaned CUDA contexts after 100s — container needs restart.
FAIL — Aborting CI
```

This is the same infrastructure failure after rerunning the failed jobs. It then fast-failed the remaining H100 partitions 2/3, B200 partition 2, and H200 partition 2. Base A (including the new MXFP4 sharded-state regression), Base B, and the Base C partitions that actually ran passed. The 4-GPU MXFP4 E2E test was not in the completed H100/B200 partitions, so that validation is still blocked by the fast-fail cascade.

@Fridge003, could you please recycle/restart the B300 runner and rerun the failed jobs? If that cannot be done immediately, could a maintainer apply `bypass-fastfail` and rerun the failed jobs so the remaining H100/B200 validation can complete?

**Comment by mmangkad** [comment]
/rerun-test test/registered/models_e2e/test_gpt_oss_4gpu_mxfp4.py

**Comment by jeremyzhang866** [comment]
Thanks @mmangkad for triggering the targeted CI run. I’ll continue investigating the root cause of the 4×H100 failure and will request another CI run once I have a clearer diagnosis or a targeted change. No rerun is needed yet.

**Comment by mmangkad** [comment]
Removing run-ci label for now
